### PR TITLE
[Feat][Ascend] Add Ascend RoPE and RMS-RoPE backend

### PR DESCRIPTION
**Pull request by Big2Wheel**
## What this PR does / why we need it

Adds an optional Huawei Ascend backend for Comfy Kitchen's RoPE and RMS-RoPE APIs.

Supported public APIs:

- `apply_rope`, `apply_rope1`, `apply_rope_split_half`, and `apply_rope_split_half1`
- `rms_rope`, `rms_rope1`, `rms_rope_split_half`, and `rms_rope_split_half1`
- The corresponding in-place variants of all APIs above

The backend uses `torch_npu.npu_rotary_mul` for rotary embeddings and `torch_npu.npu_rms_norm` for fused RMS normalization. It supports FP16/BF16, BNSD/BSND layouts, interleaved and split-half rotary conventions, packed-QKV strided views with a contiguous final dimension, GQA-style head counts, and partial rotary dimensions where exposed by the public API.

Capability checks reject unsupported shapes, layouts, dtypes, and broadcast patterns before dispatch, allowing the registry to select another capable backend. Runtime operator errors are not silently swallowed.

## Relationship to #151 and #153

This is a stacked PR that currently depends on #151, which introduces the generic Ascend backend registration, availability detection, and baseline quantization support. The PR-specific commits to review on top of #151 are the RoPE implementation and its two capability-validation follow-ups.

#153 is a sibling feature PR, not a dependency of this PR. Both #153 and this PR branch directly from #151:

```text
#151 Ascend backend foundation
├── #153 Ascend INT8 linear
└── this PR: Ascend RoPE / RMS-RoPE
```

After #151 merges, this branch will be rebased onto `main`; the PR diff will then contain only the RoPE/RMS-RoPE changes. No INT8-linear changes from #153 are included here.

## Does this PR introduce any user-facing change?

Yes. On supported Ascend NPU installations, the registry can select the new Ascend implementation for compatible RoPE and RMS-RoPE calls. Other devices and unsupported Ascend call patterns retain the existing backend fallback behavior.

## How was this patch tested?

Hardware: Huawei Ascend 910B3, restricted to one NPU with `ASCEND_RT_VISIBLE_DEVICES=7`.

### Tests

```text
pytest -q tests/test_ascend_backend.py tests/test_ascend_rope.py tests/test_backends.py\n109 passed after rebasing onto comfy-kitchen v0.2.33

pytest -q tests/test_ascend_backend.py tests/test_ascend_rope.py \
  tests/test_rope.py tests/test_rms_rope.py tests/test_backends.py \
  tests/test_constraints.py
319 passed, 404 skipped
```

Coverage includes FP16/BF16, BSND/BNSD, pair/single APIs, interleaved/split-half conventions, in-place/out-of-place behavior, packed QKV views, partial rotary dimensions, GQA head counts, input preservation, alias rejection, unsupported-call fallback, and automatic backend selection.

`ruff check` and `git diff --check` also pass for all changed Python files.

### Profiler verification

A torch profiler trace confirms that the implementation reaches the Ascend operators:

```text
npu::npu_rms_norm
aclnnRmsNorm
npu::npu_rotary_mul
aclnnRotaryPositionEmbeddingV2
```

No layout-induced `contiguous` operation appears in the profiled supported path.

### Operator benchmark

Each measurement used 10 warm-up iterations and 30 synchronized measured iterations. Times include coefficient conversion and required dtype semantics.

| Workload shape | Operation | Eager mean | Ascend mean | Improvement |
| --- | --- | ---: | ---: | ---: |
| Z-Image `[1, 4096, 30, 128]` | RoPE | 1.206 ms | 0.551 ms | 54.2% |
| Z-Image `[1, 4096, 30, 128]` | RMS-RoPE | 1.404 ms | 0.638 ms | 54.5% |
| Wan `[1, 12288, 16, 128]` | RoPE | 4.023 ms | 0.614 ms | 84.8% |
| Wan `[1, 12288, 16, 128]` | RMS-RoPE | 4.505 ms | 0.614 ms | 86.3% |

### End-to-end workflow validation

Z-Image BF16 showed the expected operator dispatch (272 RMS-RoPE calls per request), but no stable end-to-end improvement, so no image-workflow speedup is claimed.

For a fixed-seed Wan2.2 14B text-to-video workflow at 640x640 and 81 frames (320 RoPE calls per request):

| Version | Hot-run mean | Median | Std. dev. |
| --- | ---: | ---: | ---: |
| #151 baseline | 96.34 s | 95.97 s | 1.44 s |
| This PR | 91.79 s | 91.71 s | 0.39 s |

This is a 4.73% mean end-to-end reduction. Decoding both generated videos and comparing all 81 frames produced zero differing pixel values (`max_diff = 0`).

### CODE REVIEW

File: comfy_kitchen/backends/ascend/__init__.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    import torch_npu
+
+    if not torch.npu.is_available():
+        _ASCEND_ERROR = "torch-npu is installed, but no Huawei Ascend device is available"
+    elif not hasattr(torch_npu, "npu_dynamic_quant"):
+        _ASCEND_ERROR = "torch-npu does not provide npu_dynamic_quant"
+    elif not hasattr(torch_npu, "npu_quantize"):
+        _ASCEND_ERROR = "torch-npu does not provide npu_quantize"
+    else:
+        _ASCEND_AVAILABLE = True
```
[the remark this answers is not in the document]
Comment (Big2Wheel) [reply comfy_kitchen/backends/ascend/__init__.py]: Good catch. Fixed in dcf6e20 by checking the registered torch operator schemas before exposing capabilities. The base Ascend backend now requires `npu_quantize` to provide `div_mode`, while the RoPE capability is independently gated on `npu_rotary_mul` providing `rotary_mode`; unsupported older torch-npu versions therefore fall back before dispatch. `npu_dynamic_quant` does not take either parameter, so it remains presence-checked only. Added coverage for both required parameters and the missing-schema case. Verified on Ascend 910B3: 82 targeted tests passed; the related suite completed with 317 passed and 404 skipped.

File: comfy_kitchen/backends/ascend/__init__.py [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+    _ASCEND_ERROR = f"torch-npu is not installed: {exc}"
+except Exception as exc:
+    _ASCEND_ERROR = f"torch-npu initialization failed: {exc}"
+
+
+_ASCEND_ROPE_AVAILABLE = (
+    _ASCEND_AVAILABLE
+    and hasattr(torch_npu, "npu_rotary_mul")
+    and _operator_has_parameter(torch_npu.npu_rotary_mul, "rotary_mode")
+)
```
[the remark this answers is not in the document]
Comment (Big2Wheel) [reply comfy_kitchen/backends/ascend/__init__.py:65]: Agreed. Fixed in 4b04d2f by separating `_ASCEND_DEVICE_AVAILABLE`, `_ASCEND_QUANT_AVAILABLE`, `_ASCEND_ROPE_AVAILABLE`, and `_ASCEND_RMS_ROPE_AVAILABLE`. The backend now registers whenever an Ascend device is available; quantization constraints are added only when the required quant operators/schema are compatible, while RoPE depends only on `npu_rotary_mul` plus `rotary_mode`, and RMS-RoPE additionally requires `npu_rms_norm`. This preserves supported RoPE on torch-npu builds whose quantization schema is older.

File: tests/test_ascend_rope.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+@pytest.fixture
+def ascend_device():
+    torch.npu.set_device("npu:0")
+    return torch.device("npu:0")
+
+
+def test_required_operator_parameters_are_available():
+    assert ascend_backend._operator_has_parameter(torch_npu.npu_quantize, "div_mode")
+    assert ascend_backend._operator_has_parameter(torch_npu.npu_rotary_mul, "rotary_mode")
```
[the remark this answers is not in the document]
Comment (Big2Wheel) [reply tests/test_ascend_rope.py:37]: Agreed. Fixed in 4b04d2f. The installed quantization and rotary schema assertions are now separate tests, each guarded by its corresponding capability flag before directly accessing the operator. Missing/older schema behavior is covered independently with a synthetic operator, including a no-schema object. Validation on Ascend 910B3: 109 targeted backend tests passed; the complete related suite passed with 319 tests and 404 expected skips.
